### PR TITLE
feat: core functionality added

### PR DESCRIPTION
**Pull request by Ceres6**
Core functionality added with an `onRoute` hook to add constraints to all routes declared after registration on the same or descendent scope

resolves fastify/fastify#3615

### CODE REVIEW

File: .nyc_output/039a84fc-6305-4919-b79b-b9f1c64c1724.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (simoneb) [inline .nyc_output/039a84fc-6305-4919-b79b-b9f1c64c1724.json:1]: none of the files in this folder should be committed

Comment (Ceres6) [reply .nyc_output/039a84fc-6305-4919-b79b-b9f1c64c1724.json]: My bad, fixed

File: README.md [hunk 2]
```diff
@@ -1,12 +1,45 @@
-![CI](https://github.com/nearform/bench-template/actions/workflows/ci.yml/badge.svg?event=push)
+![CI](https://github.com/nearform/fastify-constraints/actions/workflows/ci.yml/badge.svg?event=push)
 
-# Bench Template
+Fastify plugin to add constraints to multiple routes
```
Comment (simoneb) [inline README.md:3]: include a h1 with the name of the plugin, before the description, e.g. # fastify-constraints

File: README.md [hunk 3]
```diff
@@ -1,12 +1,45 @@
...
-- notifications about commits waiting to be released thanks to ["notify release" GitHub action](https://github.com/nearform/github-action-notify-release)
-- PRs' linked issues check with ["check linked issues" GitHub action](https://github.com/nearform/github-action-check-linked-issues)
-- Continuous Integration GitHub workflow
+```bash
+npm i fastify-constraints
+```
+
+## Usage
+
+Register fastifyConstraints as a Fastify plugin.
```
Comment (simoneb) [inline README.md:13]: let's name the plugin consistently, it's called fastify-constraints. Also, consider wrapping it in backticks to format it as code

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+type FastifyConstraintsOptions = {
+  constraints: { [name: string]: any}
+}
+
+declare const fastifyConstraints: FastifyPluginAsync<FastifyConstraintsOptions>
+
+export { fastifyConstraints, FastifyConstraintsOptions }
+
+export default fastifyConstraints
```
Comment (simoneb) [inline index.d.ts:9]: missing new line at end of file

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+type FastifyConstraintsOptions = {
+  constraints: { [name: string]: any}
```
Comment (simoneb) [inline index.d.ts:2]: formatting (missing space before closing brace)

File: index.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import fp from 'fastify-plugin'
+
+function fastifyConstraints(fastify, options, next) {
+  fastify.addHook('onRoute', function hook(routeOptions) {
+    routeOptions.constraints = {
+      ...options.constraints,
+      ...(routeOptions?.constraints || {})
```
Comment (simoneb) [inline index.js:7]: || {} is unnecessary. destructuring a falsy value is a noop

File: test/index.test.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
+import { test } from 'tap'
+import Fastify from 'fastify'
+import fastifyConstraints from '../index.js'
+
+test('it should add constraints to all routes', async t => {
+  t.plan(2)
```
Comment (simoneb) [inline test/index.test.js:6]: this is unnecessary, remove

File: test/index.test.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+  const fastify = Fastify()
+  await fastify.register(fastifyConstraints, {
+    constraints: { version: '1.0.0' }
+  })
+
+  fastify.get('/', () => {})
+  fastify.post('/some-route', () => {})
+
+  t.equal(
```
Comment (simoneb) [inline test/index.test.js:16]: nit: instead of comparing with true, you can simply use t.ok

File: test/index.test.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  t.plan(1)
+
+  const fastify = Fastify()
+  await fastify.register(fastifyConstraints, {
+    constraints: { version: '1.0.0' }
+  })
+  await fastify.register((instance, opts, done) => {
+    fastify.get('/', () => {})
+
+    done()
```
Comment (simoneb) [inline test/index.test.js:88]: I guess there's no need for the callback based registration syntax. Use an async function instead

**Review by simoneb — COMMENTED** [review]
Overall very good in the sense that the plugin functionality seems correct and complete 🚀 

I've added a few comments around mostly minor things and style

**Review by climba03003 — COMMENTED** [review]
It would be better to allow 

```js
fastify.register(function(fastify, option) {
  fastify.get('/', () => {})
}, { constraints: { foo: 'bar' } })
```

The plugin would not take any option and create a decorator to store the `constraints`.

```js
consy symbol = Symbol.for('kConstraints')
fastify.decorate(symbol, null)

fastify.addHook('onRegister', function(instance, options) {
  instance[symbol] = { ...instance[symbol], ...options.constraints }
})

fastify.addHook('onRoute', function(routeOptions) {
  routeOptions.constraints = { ...this[symbol], ...routeOptions.constraints }
})
```

File: index.d.ts (hunk 5 above)
Comment (Ceres6) [reply index.d.ts:2]: I'll be removing the file as the new implementation won't take any options

**Comment by Ceres6** [comment]
> It would be better to allow
> 
> ```js
> fastify.register(function(fastify, option) {
>   fastify.get('/', () => {})
> }, { constraints: { foo: 'bar' } })
> ```
> 
> The plugin would not take any option and create a property to store the `constraints`.
> 
> ```js
> const symbol = Symbol.for('kConstraints')
> fastify.decorate(symbol, null)
> 
> fastify.addHook('onRegister', function(instance, options) {
>   if(options.constraints) {
>     instance[symbol] = { ...instance[symbol], ...options.constraints }
>   }
> })
> 
> fastify.addHook('onRoute', function(routeOptions) {
>   if(this[symbol]) {
>     routeOptions.constraints = { ...this[symbol], ...routeOptions.constraints }
>   }
> })
> ```

Hi @climba03003  Thanks for pointing it out I was already working on making it that way. It should work that way now. Let me know if you have any other comments

File: index.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import fp from 'fastify-plugin'
+
+function fastifyConstraints(fastify, options, next) {
+  fastify.addHook('onRegister', (instance, opts) => {
+    if (!opts?.constraints) {
+      next()
```
Comment (climba03003) [inline index.js:6]: There are no `next` function inside.

```suggestion
```

File: index.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+      return
+    }
+
+    instance.addHook('onRoute', function hook(routeOptions) {
```
Comment (climba03003) [inline index.js:10]: Not sure if it is good or not. It may add a lot of `onRoute` function.

Comment (Ceres6) [reply index.js:9]: I see your point, we would have as many on route hooks as plugins with constraints. I think your way would be more efficient. Let me change that.

File: examples/imported-routes.cjs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+const Fastify = require('fastify')
```
Comment (simoneb) [inline examples/imported-routes.cjs:1]: let's remove the cjs ones

**Review by simoneb — APPROVED** [review]
LGTM 🚀

**Pull request merged**